internal/tekton-bundle/tekton-bundle.go

```diff
@@ -0,0 +1,179 @@
+package tekton_bundle
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+
+	"path/filepath"
+
+	openshiftconfigv1 "github.com/openshift/api/config/v1"
+	pipeline "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	yamlv2 "gopkg.in/yaml.v2"
+	v1 "k8s.io/api/core/v1"
+	rbac "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	tektonPipelinesKubernetesBundleDir = "/data/tekton-pipelines/kubernetes/"
+	tektonPipelinesOKDBundleDir        = "/data/tekton-pipelines/okd/"
+)
+
+var (
+	pipelineKindString = "Pipeline"
+	serviceAccountKind = rbac.ServiceAccountKind
+	roleBindingKind    = "RoleBinding"
+	clusterRoleKind    = "ClusterRole"
+	configMapKind      = "ConfigMap"
+)
+
+type Bundle struct {
+	ServiceAccounts []v1.ServiceAccount
+	RoleBindings    []rbac.RoleBinding
+	ClusterRoles    []rbac.ClusterRole
+	Pipelines       []pipeline.Pipeline
+	ConfigMaps      []v1.ConfigMap
+}
+
+func ReadPipelineBundle(cl client.Reader, ctx context.Context) (*Bundle, error) {
+	isOpenshift, err := runningOnOpenshift(cl, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	path := getPipelineBundlePath(isOpenshift)
+	files, err := readFolder(path)
```

Comment:
Hi, how do you think we can achieve it? Generators in Go aren't supported, from what I see it can be achieved via Goroutines/Channels. This is what you mean?

Reply:
Simply merge the functions `readFolder()` and `decodeObjectsFromFiles()`. 
The outer loops in both functions can be combined, so that each file is processed after it is loaded, and then the next file is loaded and porcessed.

Reply:
I wish I could do it, but that `decodeObjectsFromFiles()` is used in both different functions:
```
func ReadTasksBundle(isOpenshift bool) (*Bundle, error) {
	var files [][]byte
	path := getTasksBundlePath(isOpenshift)
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	files = append(files, data)

	tektonObjs, err := decodeObjectsFromFiles(files)
	if err != nil {
		return nil, err
	}

	return tektonObjs, nil
}

func ReadPipelineBundle() (*Bundle, error) {
	path := getPipelineBundlePath()
	files, err := readFolder(path)
	if err != nil {
		return nil, err
	}

	tektonObjs, err := decodeObjectsFromFiles(files)
	if err != nil {
		return nil, err
	}

	return tektonObjs, nil
}
```
I'll take a look.

Reply:
You could, for example, change the argument of `decodeObjectsFromFiles()` function to be a list of filenames, and then inside it would open and load the files one by one.

Reply:
Done.